<a href="https://colab.research.google.com/github/AGON002/python-assignment/blob/main/EDA_Assg_NYC_Taxi_Starter.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **New York City Yellow Taxi Data**

## Objective
In this case study you will be learning exploratory data analysis (EDA) with the help of a dataset on yellow taxi rides in New York City. This will enable you to understand why EDA is an important step in the process of data science and machine learning.

## **Problem Statement**
As an analyst at an upcoming taxi operation in NYC, you are tasked to use the 2023 taxi trip data to uncover insights that could help optimise taxi operations. The goal is to analyse patterns in the data that can inform strategic decisions to improve service efficiency, maximise revenue, and enhance passenger experience.

## Tasks
You need to perform the following steps for successfully completing this assignment:
1. Data Loading
2. Data Cleaning
3. Exploratory Analysis: Bivariate and Multivariate
4. Creating Visualisations to Support the Analysis
5. Deriving Insights and Stating Conclusions

---

**NOTE:** The marks given along with headings and sub-headings are cumulative marks for those particular headings/sub-headings.<br>

The actual marks for each task are specified within the tasks themselves.

For example, marks given with heading *2* or sub-heading *2.1* are the cumulative marks, for your reference only. <br>

The marks you will receive for completing tasks are given with the tasks.

Suppose the marks for two tasks are: 3 marks for 2.1.1 and 2 marks for 3.2.2, or
* 2.1.1 [3 marks]
* 3.2.2 [2 marks]

then, you will earn 3 marks for completing task 2.1.1 and 2 marks for completing task 3.2.2.


---

## Data Understanding
The yellow taxi trip records include fields capturing pick-up and drop-off dates/times, pick-up and drop-off locations, trip distances, itemized fares, rate types, payment types, and driver-reported passenger counts.

The data is stored in Parquet format (*.parquet*). The dataset is from 2009 to 2024. However, for this assignment, we will only be using the data from 2023.

The data for each month is present in a different parquet file. You will get twelve files for each of the months in 2023.

The data was collected and provided to the NYC Taxi and Limousine Commission (TLC) by technology providers like vendors and taxi hailing apps. <br>

You can find the link to the TLC trip records page here: https://www.nyc.gov/site/tlc/about/tlc-trip-record-data.page

###  Data Description
You can find the data description here: [Data Dictionary](https://www.nyc.gov/assets/tlc/downloads/pdf/data_dictionary_trip_records_yellow.pdf)

**Trip Records**



|Field Name       |description |
|:----------------|:-----------|
| VendorID | A code indicating the TPEP provider that provided the record. <br> 1= Creative Mobile Technologies, LLC; <br> 2= VeriFone Inc. |
| tpep_pickup_datetime | The date and time when the meter was engaged.  |
| tpep_dropoff_datetime | The date and time when the meter was disengaged.   |
| Passenger_count | The number of passengers in the vehicle. <br> This is a driver-entered value. |
| Trip_distance | The elapsed trip distance in miles reported by the taximeter. |
| PULocationID | TLC Taxi Zone in which the taximeter was engaged |
| DOLocationID | TLC Taxi Zone in which the taximeter was disengaged |
|RateCodeID |The final rate code in effect at the end of the trip.<br> 1 = Standard rate <br> 2 = JFK <br> 3 = Newark <br>4 = Nassau or Westchester <br>5 = Negotiated fare <br>6 = Group ride |
|Store_and_fwd_flag |This flag indicates whether the trip record was held in vehicle memory before sending to the vendor, aka “store and forward,” because the vehicle did not have a connection to the server.  <br>Y= store and forward trip <br>N= not a store and forward trip |
|Payment_type| A numeric code signifying how the passenger paid for the trip. <br> 1 = Credit card <br>2 = Cash <br>3 = No charge <br>4 = Dispute <br>5 = Unknown <br>6 = Voided trip |
|Fare_amount| The time-and-distance fare calculated by the meter. <br>Extra Miscellaneous extras and surcharges.  Currently, this only includes the 0.50 and 1 USD rush hour and overnight charges. |
|MTA_tax |0.50 USD MTA tax that is automatically triggered based on the metered rate in use. |
|Improvement_surcharge | 0.30 USD improvement surcharge assessed trips at the flag drop. The improvement surcharge began being levied in 2015. |
|Tip_amount |Tip amount – This field is automatically populated for credit card tips. Cash tips are not included. |
| Tolls_amount | Total amount of all tolls paid in trip.  |
| total_amount | The total amount charged to passengers. Does not include cash tips. |
|Congestion_Surcharge |Total amount collected in trip for NYS congestion surcharge. |
| Airport_fee | 1.25 USD for pick up only at LaGuardia and John F. Kennedy Airports|

Although the amounts of extra charges and taxes applied are specified in the data dictionary, you will see that some cases have different values of these charges in the actual data.

**Taxi Zones**

Each of the trip records contains a field corresponding to the location of the pickup or drop-off of the trip, populated by numbers ranging from 1-263.

These numbers correspond to taxi zones, which may be downloaded as a table or map/shapefile and matched to the trip records using a join.

This is covered in more detail in later sections.

---

In [ ]:
# numpy version: 1.26.4
# pandas version: 2.2.2
# matplotlib version: 3.10.0
# seaborn version: 0.13.2
%pip install numpy==1.26.4 pandas==2.2.2 seaborn==0.13.2 matplotlib==3.10.0

Note: you may need to restart the kernel to use updated packages.


## **1** Data Preparation

<font color = red>[5 marks]</font> <br>

### Import Libraries

In [160]:
# Import warnings
import warnings


In [161]:
# Import the libraries you will be using for analysis

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns



In [162]:
# Recommended versions
# numpy version: 1.26.4
# pandas version: 2.2.2
# matplotlib version: 3.10.0
# seaborn version: 0.13.2

# Check versions
print("numpy version:", np.__version__)
print("pandas version:", pd.__version__)
print("matplotlib version:", plt.matplotlib.__version__)
print("seaborn version:", sns.__version__)

numpy version: 2.0.2
pandas version: 2.2.2
matplotlib version: 3.10.0
seaborn version: 0.13.2


### **1.1** Load the dataset
<font color = red>[5 marks]</font> <br>

You will see twelve files, one for each month.

To read parquet files with Pandas, you have to follow a similar syntax as that for CSV files.

`df = pd.read_parquet('file.parquet')`

In [218]:
# Try loading one file

from google.colab import drive
drive.mount('/content/drive')

df= pd.read_parquet('/content/drive/MyDrive/data_files/yellow_tripdata_2023-01.parquet')
df.info()

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3066766 entries, 0 to 3066765
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge 

How many rows are there? Do you think handling such a large number of rows is computationally feasible when we have to combine the data for all twelve months into one?

To handle this, we need to sample a fraction of data from each of the files. How to go about that? Think of a way to select only some portion of the data from each month's file that accurately represents the trends.

#### Sampling the Data
> One way is to take a small percentage of entries for pickup in every hour of a date. So, for all the days in a month, we can iterate through the hours and select 5% values randomly from those. Use `tpep_pickup_datetime` for this. Separate date and hour from the datetime values and then for each date, select some fraction of trips for each of the 24 hours.

To sample data, you can use the `sample()` method. Follow this syntax:

```Python
# sampled_data is an empty DF to keep appending sampled data of each hour
# hour_data is the DF of entries for an hour 'X' on a date 'Y'

sample = hour_data.sample(frac = 0.05, random_state = 42)
# sample 0.05 of the hour_data
# random_state is just a seed for sampling, you can define it yourself

sampled_data = pd.concat([sampled_data, sample]) # adding data for this hour to the DF
```

This *sampled_data* will contain 5% values selected at random from each hour.

Note that the code given above is only the part that will be used for sampling and not the complete code required for sampling and combining the data files.

Keep in mind that you sample by date AND hour, not just hour. (Why?)

---

**1.1.1** <font color = red>[5 marks]</font> <br>
Figure out how to sample and combine the files.

**Note:** It is not mandatory to use the method specified above. While sampling, you only need to make sure that your sampled data represents the overall data of all the months accurately.

In [219]:
# Sample the data
# It is recommmended to not load all the files at once to avoid memory overload

In [229]:
# Take a small percentage of entries from each hour of every date.
# Iterating through the monthly data:
#   read a month file -> day -> hour: append sampled data -> move to next hour -> move to next day after 24 hours -> move to next month file
# Create a single dataframe for the year combining all the monthly data

# Select the folder having data files
import os

# Select the folder having data files
os.chdir('/content/drive/MyDrive/data_files')

# Create a list of all the twelve files to read
file_list = os.listdir()

# initialise an empty dataframe
df = pd.DataFrame()
sample_fraction = 0.007

# iterate through the list of files and sample one by one:
for file_name in file_list:
    try:
        # file path for the current file
        file_path = os.path.join(os.getcwd(), file_name)

        # Reading the current file
        df_each = pd.read_parquet(file_path)
        df_each['date'] = df_each['tpep_pickup_datetime'].dt.date
        df_each['hour'] = df_each['tpep_pickup_datetime'].dt.hour

        # We will store the sampled data for the current date in this df by appending the sampled data from each hour to this
        # After completing iteration through each date, we will append this data to the final dataframe.
        sampled_data = pd.DataFrame()

        sampled = (
          df_each.groupby(['date', 'hour'], group_keys=False)
          .apply(lambda x: x.sample(frac=sample_fraction, random_state=42))
        )

        df = pd.concat([df, sampled])
    except Exception as e:
        print(f"Error reading file {file_name}: {e}")

<ipython-input-229-c64d31029aa7>:36: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  .apply(lambda x: x.sample(frac=sample_fraction, random_state=42))
<ipython-input-229-c64d31029aa7>:36: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  .apply(lambda x: x.sample(frac=sample_fraction, random_state=42))
<ipython-input-229-c64d31029aa7>:36: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns

Error reading file Taxi_data.csv: Could not open Parquet input source '<Buffer>': Parquet magic bytes not found in footer. Either the file is corrupted or this is not a parquet file.


<ipython-input-229-c64d31029aa7>:36: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  .apply(lambda x: x.sample(frac=sample_fraction, random_state=42))


After combining the data files into one DataFrame, convert the new DataFrame to a CSV or parquet file and store it to use directly.

Ideally, you can try keeping the total entries to around 250,000 to 300,000.

In [230]:
# Store the df in csv/parquet
# df.to_parquet('')
df.to_csv('Taxi_data.csv')


## **2** Data Cleaning
<font color = red>[30 marks]</font> <br>

Now we can load the new data directly.

In [231]:
# Load the new data file

df= pd.read_csv('Taxi_data.csv')

In [232]:
# df.head()
df.head()

Unnamed: 0  VendorID tpep_pickup_datetime tpep_dropoff_datetime  \
0        1164         2  2023-01-01 00:10:30   2023-01-01 00:11:49   
1        1868         2  2023-01-01 00:49:02   2023-01-01 00:55:15   
2        3106         1  2023-01-01 00:47:17   2023-01-01 01:07:01   
3         808         2  2023-01-01 00:06:02   2023-01-01 00:31:38   
4         169         2  2023-01-01 00:02:19   2023-01-01 00:30:49   

   passenger_count  trip_distance  RatecodeID store_and_fwd_flag  \
0              1.0           0.49         1.0                  N   
1              1.0           0.75         1.0                  N   
2              2.0           2.90         1.0                  N   
3              1.0           2.50         1.0                  N   
4              1.0          20.37         2.0                  N   

   PULocationID  DOLocationID  ...  mta_tax  tip_amount  tolls_amount  \
0           239           238  ...      0.5        1.88          0.00   
1            45           148  ...      0.5        0.00          0.00   
2           142           170  ...      0.5        4.80          0.00   
3            43           161  ...      0.5        7.25          0.00   
4           132           140  ...      0.5       12.00          6.55   

   improvement_surcharge  total_amount  congestion_surcharge  airport_fee  \
0                    1.0         11.28                   2.5          0.0   
1                    1.0         12.90                   2.5          0.0   
2                    1.0         28.90                   2.5          0.0   
3                    1.0         36.25                   2.5          0.0   
4                    1.0         92.55                   2.5          0.0   

         date  hour  Airport_fee  
0  2023-01-01     0          NaN  
1  2023-01-01     0          NaN  
2  2023-01-01     0          NaN  
3  2023-01-01     0          NaN  
4  2023-01-01     0          NaN  

[5 rows x 23 columns]

In [233]:
# df.info()
df.info()




<class 'pandas.core.frame.DataFrame'>
RangeIndex: 268150 entries, 0 to 268149
Data columns (total 23 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   Unnamed: 0             268150 non-null  int64  
 1   VendorID               268150 non-null  int64  
 2   tpep_pickup_datetime   268150 non-null  object 
 3   tpep_dropoff_datetime  268150 non-null  object 
 4   passenger_count        259263 non-null  float64
 5   trip_distance          268150 non-null  float64
 6   RatecodeID             259263 non-null  float64
 7   store_and_fwd_flag     259263 non-null  object 
 8   PULocationID           268150 non-null  int64  
 9   DOLocationID           268150 non-null  int64  
 10  payment_type           268150 non-null  int64  
 11  fare_amount            268150 non-null  float64
 12  extra                  268150 non-null  float64
 13  mta_tax                268150 non-null  float64
 14  tip_amount             268150 non-nu

In [234]:
df.describe()


Unnamed: 0       VendorID  passenger_count  trip_distance  \
count  2.681500e+05  268150.000000    259263.000000  268150.000000   
mean   1.604426e+06       1.736804         1.371152       4.349966   
std    9.373572e+05       0.444583         0.892247     266.246479   
min    1.400000e+01       1.000000         0.000000       0.000000   
25%    7.971672e+05       1.000000         1.000000       1.040000   
50%    1.594866e+06       2.000000         1.000000       1.780000   
75%    2.391461e+06       2.000000         1.000000       3.400000   
max    3.522222e+06       6.000000         8.000000  116439.710000   

          RatecodeID   PULocationID   DOLocationID   payment_type  \
count  259263.000000  268150.000000  268150.000000  268150.000000   
mean        1.630090     165.026459     163.957505       1.186705   
std         7.353466      64.086976      69.831952       0.556279   
min         1.000000       1.000000       1.000000       0.000000   
25%         1.000000     132.000000     114.000000       1.000000   
50%         1.000000     162.000000     162.000000       1.000000   
75%         1.000000     234.000000     234.000000       1.000000   
max        99.000000     265.000000     265.000000       4.000000   

         fare_amount          extra        mta_tax     tip_amount  \
count  268150.000000  268150.000000  268150.000000  268150.000000   
mean       19.464329       1.560645       0.485588       3.514942   
std        18.996295       1.843220       0.108607       4.112804   
min      -525.200000      -7.500000      -0.500000    -411.000000   
25%         9.300000       0.000000       0.500000       1.000000   
50%        13.500000       1.000000       0.500000       2.800000   
75%        21.900000       2.500000       0.500000       4.400000   
max       715.600000      11.750000       0.800000     126.350000   

        tolls_amount  improvement_surcharge   total_amount  \
count  268150.000000          268150.000000  268150.000000   
mean        0.585036               0.979394      28.388632   
std         2.174314               0.199302      23.687554   
min       -42.000000              -1.000000    -505.950000   
25%         0.000000               1.000000      15.900000   
50%         0.000000               1.000000      21.000000   
75%         0.000000               1.000000      30.720000   
max        72.000000               1.000000     724.150000   

       congestion_surcharge   airport_fee           hour    Airport_fee  
count         259263.000000  20975.000000  268150.000000  238288.000000  
mean               2.264197      0.111204      14.261626       0.141202  
std                0.797480      0.361067       5.813194       0.474563  
min               -2.500000     -1.250000       0.000000      -1.750000  
25%                2.500000      0.000000      11.000000       0.000000  
50%                2.500000      0.000000      15.000000       0.000000  
75%                2.500000      0.000000      19.000000       0.000000  
max                2.500000      1.250000      23.000000       1.750000

In [235]:
df.isnull().sum()


Unnamed: 0                    0
VendorID                      0
tpep_pickup_datetime          0
tpep_dropoff_datetime         0
passenger_count            8887
trip_distance                 0
RatecodeID                 8887
store_and_fwd_flag         8887
PULocationID                  0
DOLocationID                  0
payment_type                  0
fare_amount                   0
extra                         0
mta_tax                       0
tip_amount                    0
tolls_amount                  0
improvement_surcharge         0
total_amount                  0
congestion_surcharge       8887
airport_fee              247175
date                          0
hour                          0
Airport_fee               29862
dtype: int64

In [236]:
df[df.isnull().any(axis=1)]

Unnamed: 0  VendorID tpep_pickup_datetime tpep_dropoff_datetime  \
0             1164         2  2023-01-01 00:10:30   2023-01-01 00:11:49   
1             1868         2  2023-01-01 00:49:02   2023-01-01 00:55:15   
2             3106         1  2023-01-01 00:47:17   2023-01-01 01:07:01   
3              808         2  2023-01-01 00:06:02   2023-01-01 00:31:38   
4              169         2  2023-01-01 00:02:19   2023-01-01 00:30:49   
...            ...       ...                  ...                   ...   
268145     3202993         2  2023-11-30 23:34:08   2023-11-30 23:45:41   
268146     3202493         2  2023-11-30 23:28:24   2023-11-30 23:39:51   
268147     3205476         2  2023-11-30 23:14:41   2023-11-30 23:33:52   
268148     3201592         2  2023-11-30 23:02:11   2023-11-30 23:08:23   
268149     3202362         2  2023-11-30 23:28:13   2023-11-30 23:31:27   

        passenger_count  trip_distance  RatecodeID store_and_fwd_flag  \
0                   1.0           0.49         1.0                  N   
1                   1.0           0.75         1.0                  N   
2                   2.0           2.90         1.0                  N   
3                   1.0           2.50         1.0                  N   
4                   1.0          20.37         2.0                  N   
...                 ...            ...         ...                ...   
268145              1.0           1.73         1.0                  N   
268146              2.0           2.24         1.0                  N   
268147              2.0           2.50         1.0                  N   
268148              1.0           0.98         1.0                  N   
268149              3.0           0.87         1.0                  N   

        PULocationID  DOLocationID  ...  mta_tax  tip_amount  tolls_amount  \
0                239           238  ...      0.5        1.88          0.00   
1                 45           148  ...      0.5        0.00          0.00   
2                142           170  ...      0.5        4.80          0.00   
3                 43           161  ...      0.5        7.25          0.00   
4                132           140  ...      0.5       12.00          6.55   
...              ...           ...  ...      ...         ...           ...   
268145            68           170  ...      0.5        0.00          0.00   
268146           263           162  ...      0.5        3.70          0.00   
268147           264           264  ...      0.5        4.18          0.00   
268148           142            48  ...      0.5        2.00          0.00   
268149            50           142  ...      0.5        3.45          0.00   

        improvement_surcharge  total_amount  congestion_surcharge  \
0                         1.0         11.28                   2.5   
1                         1.0         12.90                   2.5   
2                         1.0         28.90                   2.5   
3                         1.0         36.25                   2.5   
4                         1.0         92.55                   2.5   
...                       ...           ...                   ...   
268145                    1.0         17.80                   2.5   
268146                    1.0         22.20                   2.5   
268147                    1.0         25.08                   0.0   
268148                    1.0         14.90                   2.5   
268149                    1.0         14.95                   2.5   

        airport_fee        date  hour  Airport_fee  
0               0.0  2023-01-01     0          NaN  
1               0.0  2023-01-01     0          NaN  
2               0.0  2023-01-01     0          NaN  
3               0.0  2023-01-01     0          NaN  
4               0.0  2023-01-01     0          NaN  
...             ...         ...   ...          ...  
268145          NaN  2023-11-30    23          0.0  
268146          NaN  2023-11-30    23   

#### **2.1** Fixing Columns
<font color = red>[10 marks]</font> <br>

Fix/drop any columns as you seem necessary in the below sections

**2.1.1** <font color = red>[2 marks]</font> <br>

Fix the index and drop unnecessary columns

In [237]:
# Fix the index and drop any columns that are not needed
df.reset_index(drop=True, inplace=True)
df.drop(columns=['Unnamed: 0'], inplace=True)
df.isnull().sum()



VendorID                      0
tpep_pickup_datetime          0
tpep_dropoff_datetime         0
passenger_count            8887
trip_distance                 0
RatecodeID                 8887
store_and_fwd_flag         8887
PULocationID                  0
DOLocationID                  0
payment_type                  0
fare_amount                   0
extra                         0
mta_tax                       0
tip_amount                    0
tolls_amount                  0
improvement_surcharge         0
total_amount                  0
congestion_surcharge       8887
airport_fee              247175
date                          0
hour                          0
Airport_fee               29862
dtype: int64

**2.1.2** <font color = red>[3 marks]</font> <br>
There are two airport fee columns. This is possibly an error in naming columns. Let's see whether these can be combined into a single column.

In [238]:
# Combine the two airport fee columns
df['Airport_fee'].fillna(df['airport_fee'], inplace=True)
df.drop(columns =['airport_fee'], inplace=True)

<ipython-input-238-49f25ac9a824>:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Airport_fee'].fillna(df['airport_fee'], inplace=True)


**2.1.3** <font color = red>[5 marks]</font> <br>
Fix columns with negative (monetary) values

In [239]:
# check where values of fare amount are negative
df[df['fare_amount'] < 0]

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
84             2  2023-01-01 02:29:07   2023-01-01 02:29:24              1.0   
171            2  2023-01-01 06:36:37   2023-01-01 06:44:17              1.0   
204            2  2023-01-01 10:15:22   2023-01-01 10:15:30              1.0   
272            2  2023-01-01 13:15:52   2023-01-01 13:28:52              1.0   
570            2  2023-01-02 07:11:35   2023-01-02 07:11:43              1.0   
...          ...                  ...                   ...              ...   
267696         2  2023-11-30 15:02:21   2023-11-30 15:05:44              1.0   
267746         2  2023-11-30 16:04:13   2023-11-30 16:06:06              1.0   
267780         2  2023-11-30 17:44:21   2023-11-30 17:44:40              1.0   
267792         2  2023-11-30 17:29:21   2023-11-30 17:29:39              1.0   
267958         2  2023-11-30 20:10:42   2023-11-30 20:17:15              1.0   

        trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
84               0.00         3.0                  N           234   
171              1.84         1.0                  N           162   
204              0.00         1.0                  N            75   
272              2.47         1.0                  N           246   
570              0.00         1.0                  N           100   
...               ...         ...                ...           ...   
267696           0.61         1.0                  N           262   
267746           0.35         1.0                  N           107   
267780           0.06         1.0                  N            70   
267792           0.03         1.0                  N           142   
267958           1.32         1.0                  N            43   

        DOLocationID  payment_type  ...  extra  mta_tax  tip_amount  \
84               234             4  ...   -1.0      0.0         0.0   
171              100             4  ...    0.0     -0.5         0.0   
204               75             1  ...    0.0     -0.5         0.0   
272              125             4  ...    0.0     -0.5         0.0   
570              100             3  ...    0.0     -0.5         0.0   
...              ...           ...  ...    ...      ...         ...   
267696           236             3  ...    0.0     -0.5         0.0   
267746           107             4  ...   -2.5     -0.5         0.0   
267780            70             3  ...   -7.5     -0.5         0.0   
267792           142             3  ...   -2.5     -0.5         0.0   
267958            43             3  ...   -1.0     -0.5         0.0   

        tolls_amount  improvement_surcharge  total_amount  \
84               0.0                   -1.0         -25.0   
171              0.0                   -1.0         -14.7   
204              0.0                   -1.0          -4.5   
272              0.0                   -1.0         -18.9   
570              0.0                   -1.0          -7.0   
...              ...                    ...           ...   
267696           0.0                   -1.0          -9.8   
267746           0.0                   -1.0         -10.9   
267780           0.0                   -1.0         -12.0   
267792           0.0                   -1.0          -9.5   
267958           0.0                   -1.0         -14.3   

        congestion_surcharge        date hour  Airport_fee  
84                       0.0  2023-01-01    2          0.0  
171                     -2.5  2023-01-01    6          0.0  
204                      0.0  2023-01-01   10          0.0  
272                     -2.5  2023-01-01   13          0.0  
570                     -2.5  2023-01-02    7          0.0  
...                      ...         ...  ...          ...  
267696                  -2.5  2023-11-30   15          0.0  
267746                  -2.5  2023-11-30   16          0.0  
267780                   0.0  2023-11-30   17          0.0  
267792    

Did you notice something different in the `RatecodeID` column for above records?

In [240]:
# Analyse RatecodeID for the negative fare amounts
df[df['fare_amount'] < 0]['RatecodeID'].value_counts()

RatecodeID
1.0    2353
2.0     193
5.0      40
3.0      27
4.0      12
Name: count, dtype: int64

In [241]:
# Find which columns have negative values
neg_cols = df.select_dtypes(include=['number']).columns
negative_value_counts = (df[neg_cols] < 0).sum()
negative_value_counts[negative_value_counts > 0]



fare_amount              2668
extra                    1344
mta_tax                  2571
tip_amount                 11
tolls_amount              169
improvement_surcharge    2638
total_amount             2641
congestion_surcharge     2117
Airport_fee               353
dtype: int64

In [242]:
# fix these negative values
cols_with_negatives = [
    'fare_amount', 'extra', 'mta_tax', 'tip_amount',
    'tolls_amount', 'improvement_surcharge', 'total_amount',
    'congestion_surcharge', 'Airport_fee'
]

for col in cols_with_negatives:
    df[col] = df[col].clip(lower=0)

(df[cols_with_negatives] < 0).sum()


fare_amount              0
extra                    0
mta_tax                  0
tip_amount               0
tolls_amount             0
improvement_surcharge    0
total_amount             0
congestion_surcharge     0
Airport_fee              0
dtype: int64

### **2.2** Handling Missing Values
<font color = red>[10 marks]</font> <br>

**2.2.1**  <font color = red>[2 marks]</font> <br>
Find the proportion of missing values in each column




In [243]:
# Find the proportion of missing values in each column
df.isnull().sum()



VendorID                    0
tpep_pickup_datetime        0
tpep_dropoff_datetime       0
passenger_count          8887
trip_distance               0
RatecodeID               8887
store_and_fwd_flag       8887
PULocationID                0
DOLocationID                0
payment_type                0
fare_amount                 0
extra                       0
mta_tax                     0
tip_amount                  0
tolls_amount                0
improvement_surcharge       0
total_amount                0
congestion_surcharge     8887
date                        0
hour                        0
Airport_fee              8887
dtype: int64

**2.2.2**  <font color = red>[3 marks]</font> <br>
Handling missing values in `passenger_count`

In [244]:
# Display the rows with null values
# Impute NaN values in 'passenger_count'
mean_passenger_count = df['passenger_count'].mean().round()
df[df['passenger_count'].isnull()]
df['passenger_count'].fillna(mean_passenger_count, inplace=True)

<ipython-input-244-f6a7e1942724>:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['passenger_count'].fillna(mean_passenger_count, inplace=True)


Did you find zeroes in passenger_count? Handle these.

In [245]:
# checking if any row has passenger_count equalto zero
df['passenger_count'].value_counts()
df[df['passenger_count']==0]
df.loc[df['passenger_count'] == 0, 'passenger_count'] = mean_passenger_count

**2.2.3**  <font color = red>[2 marks]</font> <br>
Handle missing values in `RatecodeID`

In [246]:
# Fix missing values in 'RatecodeID'
mode_val = df['RatecodeID'].mode()[0]
df['RatecodeID'].fillna(mode_val, inplace=True)
df['RatecodeID']



<ipython-input-246-6f68c277a3eb>:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['RatecodeID'].fillna(mode_val, inplace=True)


0         1.0
1         1.0
2         1.0
3         1.0
4         2.0
         ... 
268145    1.0
268146    1.0
268147    1.0
268148    1.0
268149    1.0
Name: RatecodeID, Length: 268150, dtype: float64

**2.2.4**  <font color = red>[3 marks]</font> <br>
Impute NaN in `congestion_surcharge`

In [247]:
# handle null values in congestion_surcharge
df['congestion_surcharge'].fillna(0, inplace=True)

df[df['congestion_surcharge'].isnull()].value_counts()


<ipython-input-247-a34180b7f340>:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['congestion_surcharge'].fillna(0, inplace=True)


Series([], Name: count, dtype: int64)

Are there missing values in other columns? Did you find NaN values in some other set of columns? Handle those missing values below.

In [248]:
# Handle any remaining missing values

# df.isnull().sum()
df['Airport_fee'].fillna(df['Airport_fee'].mean(), inplace=True)
df['store_and_fwd_flag'].fillna('N', inplace=True)
# df.head()


<ipython-input-248-928524962db5>:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Airport_fee'].fillna(df['Airport_fee'].mean(), inplace=True)
<ipython-input-248-928524962db5>:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplac

### **2.3** Handling Outliers
<font color = red>[10 marks]</font> <br>

Before we start fixing outliers, let's perform outlier analysis.

In [252]:
# Describe the data and check if there are any potential outliers present
# Check for potential out of place values in various columns
# df.isna().sum()
# df.isnull().sum()
df.describe()

VendorID  passenger_count  trip_distance     RatecodeID  \
count  268147.000000    268147.000000  268147.000000  268147.000000   
mean        1.736801         1.373765       4.349952       1.609170   
std         0.444584         0.864686     266.247967       7.231496   
min         1.000000         1.000000       0.000000       1.000000   
25%         1.000000         1.000000       1.040000       1.000000   
50%         2.000000         1.000000       1.780000       1.000000   
75%         2.000000         1.000000       3.400000       1.000000   
max         6.000000         6.000000  116439.710000      99.000000   

        PULocationID   DOLocationID   payment_type    fare_amount  \
count  268147.000000  268147.000000  268147.000000  268147.000000   
mean      165.026459     163.957374       1.186704      19.668181   
std        64.087084      69.832195       0.556279      18.470748   
min         1.000000       1.000000       0.000000       0.000000   
25%       132.000000     114.000000       1.000000       9.300000   
50%       162.000000     162.000000       1.000000      13.500000   
75%       234.000000     234.000000       1.000000      21.900000   
max       265.000000     265.000000       4.000000     715.600000   

               extra        mta_tax     tip_amount   tolls_amount  \
count  268147.000000  268147.000000  268147.000000  268147.000000   
mean        1.570828       0.490384       3.516496       0.590095   
std         1.825942       0.068695       4.033967       2.159835   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.500000       1.000000       0.000000   
50%         1.000000       0.500000       2.800000       0.000000   
75%         2.500000       0.500000       4.400000       0.000000   
max        11.750000       0.800000     126.350000      72.000000   

       improvement_surcharge   total_amount  congestion_surcharge  \
count          268147.000000  268147.000000         268147.000000   
mean                0.989229      28.644123              2.208900   
std                 0.102591      23.057358              0.801881   
min                 0.000000       0.000000              0.000000   
25%                 1.000000      15.900000              2.500000   
50%                 1.000000      21.000000              2.500000   
75%                 1.000000      30.720000              2.500000   
max                 1.000000     724.150000              2.500000   

                hour    Airport_fee  
count  268147.000000  268147.000000  
mean       14.261733       0.141017  
std         5.813137       0.454060  
min         0.000000       0.000000  
25%        11.000000       0.000000  
50%        15.000000       0.000000  
75%        19.000000       0.000000  
max        23.000000       1.750000

**2.3.1**  <font color = red>[10 marks]</font> <br>
Based on the above analysis, it seems that some of the outliers are present due to errors in registering the trips. Fix the outliers.

Some points you can look for:
- Entries where `trip_distance` is nearly 0 and `fare_amount` is more than 300
- Entries where `trip_distance` and `fare_amount` are 0 but the pickup and dropoff zones are different (both distance and fare should not be zero for different zones)
- Entries where `trip_distance` is more than 250  miles.
- Entries where `payment_type` is 0 (there is no payment_type 0 defined in the data dictionary)

These are just some suggestions. You can handle outliers in any way you wish, using the insights from above outlier analysis.

How will you fix each of these values? Which ones will you drop and which ones will you replace?

First, let us remove 7+ passenger counts as there are very less instances.

In [250]:
# remove passenger_count > 6
df = df[df['passenger_count'] <= 6]
df = df[df['trip_distance'] >= 1 and df['fare_amount'] < 300 and df['PULocationID'] != df['DOLocationID']]
df = df[df['trip_distance'] != 0 and df['fare_amount'] != 0]
df = df[df['trip_distance'] < 250]
df = df[df['payment_type'] != 0]



df['passenger_count'].value_counts()

passenger_count
1.0    207742
2.0     39572
3.0      9674
4.0      5562
5.0      3367
6.0      2230
Name: count, dtype: int64

In [ ]:
# Continue with outlier handling



In [ ]:
# Do any columns need standardising?



## **3** Exploratory Data Analysis
<font color = red>[90 marks]</font> <br>

In [28]:
df.columns.tolist()

['VendorID',
 'tpep_pickup_datetime',
 'tpep_dropoff_datetime',
 'passenger_count',
 'trip_distance',
 'RatecodeID',
 'store_and_fwd_flag',
 'PULocationID',
 'DOLocationID',
 'payment_type',
 'fare_amount',
 'extra',
 'mta_tax',
 'tip_amount',
 'tolls_amount',
 'improvement_surcharge',
 'total_amount',
 'congestion_surcharge',
 'airport_fee',
 'date',
 'hour',
 'Airport_fee']

#### **3.1** General EDA: Finding Patterns and Trends
<font color = red>[40 marks]</font> <br>

**3.1.1** <font color = red>[3 marks]</font> <br>
Categorise the varaibles into Numerical or Categorical.
* `VendorID`:
* `tpep_pickup_datetime`:
* `tpep_dropoff_datetime`:
* `passenger_count`:
* `trip_distance`:
* `RatecodeID`:
* `PULocationID`:
* `DOLocationID`:
* `payment_type`:
* `pickup_hour`:
* `trip_duration`:


The following monetary parameters belong in the same category, is it categorical or numerical?


* `fare_amount`
* `extra`
* `mta_tax`
* `tip_amount`
* `tolls_amount`
* `improvement_surcharge`
* `total_amount`
* `congestion_surcharge`
* `airport_fee`

##### Temporal Analysis

**3.1.2** <font color = red>[5 marks]</font> <br>
Analyse the distribution of taxi pickups by hours, days of the week, and months.

In [29]:
# Find and show the hourly trends in taxi pickups



In [30]:
# Find and show the daily trends in taxi pickups (days of the week)



In [31]:
# Show the monthly trends in pickups



##### Financial Analysis

Take a look at the financial parameters like `fare_amount`, `tip_amount`, `total_amount`, and also `trip_distance`. Do these contain zero/negative values?

In [32]:
# Analyse the above parameters



Do you think it is beneficial to create a copy DataFrame leaving out the zero values from these?

**3.1.3** <font color = red>[2 marks]</font> <br>
Filter out the zero values from the above columns.

**Note:** The distance might be 0 in cases where pickup and drop is in the same zone. Do you think it is suitable to drop such cases of zero distance?

In [33]:
# Create a df with non zero entries for the selected parameters.



**3.1.4** <font color = red>[3 marks]</font> <br>
Analyse the monthly revenue (`total_amount`) trend

In [34]:
# Group data by month and analyse monthly revenue



**3.1.5** <font color = red>[3 marks]</font> <br>
Show the proportion of each quarter of the year in the revenue

In [35]:
# Calculate proportion of each quarter



**3.1.6** <font color = red>[3 marks]</font> <br>
Visualise the relationship between `trip_distance` and `fare_amount`. Also find the correlation value for these two.

**Hint:** You can leave out the trips with trip_distance = 0

In [36]:
# Show how trip fare is affected by distance



**3.1.7** <font color = red>[5 marks]</font> <br>
Find and visualise the correlation between:
1. `fare_amount` and trip duration (pickup time to dropoff time)
2. `fare_amount` and `passenger_count`
3. `tip_amount` and `trip_distance`

In [37]:
# Show relationship between fare and trip duration



In [38]:
# Show relationship between fare and number of passengers



In [39]:
# Show relationship between tip and trip distance



**3.1.8** <font color = red>[3 marks]</font> <br>
Analyse the distribution of different payment types (`payment_type`)

In [40]:
# Analyse the distribution of different payment types (payment_type).




- 1= Credit card
- 2= Cash
- 3= No charge
- 4= Dispute



##### Geographical Analysis

For this, you have to use the *taxi_zones.shp* file from the *taxi_zones* folder.

There would be multiple files inside the folder (such as *.shx, .sbx, .sbn* etc). You do not need to import/read any of the files other than the shapefile, *taxi_zones.shp*.

Do not change any folder structure - all the files need to be present inside the folder for it to work.

The folder structure should look like this:
```
Taxi Zones
|- taxi_zones.shp.xml
|- taxi_zones.prj
|- taxi_zones.sbn
|- taxi_zones.shp
|- taxi_zones.dbf
|- taxi_zones.shx
|- taxi_zones.sbx

 ```

 You only need to read the `taxi_zones.shp` file. The *shp* file will utilise the other files by itself.

We will use the *GeoPandas* library for geopgraphical analysis
```
import geopandas as gpd
```

More about geopandas and shapefiles: [About](https://geopandas.org/en/stable/about.html)


Reading the shapefile is very similar to *Pandas*. Use `gpd.read_file()` function to load the data (*taxi_zones.shp*) as a GeoDataFrame. Documentation: [Reading and Writing Files](https://geopandas.org/en/stable/docs/user_guide/io.html)

In [41]:
# !pip install geopandas

**3.1.9** <font color = red>[2 marks]</font> <br>
Load the shapefile and display it.

In [42]:
# import geopandas as gpd


# Read the shapefile using geopandas
zones = # read the .shp file using gpd
zones.head()

SyntaxError: invalid syntax (<ipython-input-42-92b6d5fb067f>, line 5)

Now, if you look at the DataFrame created, you will see columns like: `OBJECTID`,`Shape_Leng`, `Shape_Area`, `zone`, `LocationID`, `borough`, `geometry`.
<br><br>

Now, the `locationID` here is also what we are using to mark pickup and drop zones in the trip records.

The geometric parameters like shape length, shape area and geometry are used to plot the zones on a map.

This can be easily done using the `plot()` method.

In [ ]:
# print(zones.info())
# zones.plot()

Now, you have to merge the trip records and zones data using the location IDs.



**3.1.10** <font color = red>[3 marks]</font> <br>
Merge the zones data into trip data using the `locationID` and `PULocationID` columns.

In [ ]:
# Merge zones and trip records using locationID and PULocationID



**3.1.11** <font color = red>[3 marks]</font> <br>
Group data by location IDs to find the total number of trips per location ID

In [ ]:
# Group data by location and calculate the number of trips



**3.1.12** <font color = red>[2 marks]</font> <br>
Now, use the grouped data to add number of trips to the GeoDataFrame.

We will use this to plot a map of zones showing total trips per zone.

In [ ]:
# Merge trip counts back to the zones GeoDataFrame




The next step is creating a color map (choropleth map) showing zones by the number of trips taken.

Again, you can use the `zones.plot()` method for this. [Plot Method GPD](https://geopandas.org/en/stable/docs/reference/api/geopandas.GeoDataFrame.plot.html#geopandas.GeoDataFrame.plot)

But first, you need to define the figure and axis for the plot.

`fig, ax = plt.subplots(1, 1, figsize = (12, 10))`

This function creates a figure (fig) and a single subplot (ax)

---

After setting up the figure and axis, we can proceed to plot the GeoDataFrame on this axis. This is done in the next step where we use the plot method of the GeoDataFrame.

You can define the following parameters in the `zones.plot()` method:
```
column = '',
ax = ax,
legend = True,
legend_kwds = {'label': "label", 'orientation': "<horizontal/vertical>"}
```

To display the plot, use `plt.show()`.

**3.1.13** <font color = red>[3 marks]</font> <br>
Plot a color-coded map showing zone-wise trips

In [ ]:
# Define figure and axis


# Plot the map and display it



In [ ]:
# can you try displaying the zones DF sorted by the number of trips?



Here we have completed the temporal, financial and geographical analysis on the trip records.

**Compile your findings from general analysis below:**

You can consider the following points:

* Busiest hours, days and months
* Trends in revenue collected
* Trends in quarterly revenue
* How fare depends on trip distance, trip duration and passenger counts
* How tip amount depends on trip distance
* Busiest zones


#### **3.2** Detailed EDA: Insights and Strategies
<font color = red>[50 marks]</font> <br>

Having performed basic analyses for finding trends and patterns, we will now move on to some detailed analysis focussed on operational efficiency, pricing strategies, and customer experience.

##### Operational Efficiency

Analyze variations by time of day and location to identify bottlenecks or inefficiencies in routes

**3.2.1** <font color = red>[3 marks]</font> <br>
Identify slow routes by calculating the average time taken by cabs to get from one zone to another at different hours of the day.

Speed on a route *X* for hour *Y* = (*distance of the route X / average trip duration for hour Y*)

In [ ]:
# Find routes which have the slowest speeds at different times of the day



How does identifying high-traffic, high-demand routes help us?

**3.2.2** <font color = red>[3 marks]</font> <br>
Calculate the number of trips at each hour of the day and visualise them. Find the busiest hour and show the number of trips for that hour.

In [ ]:
# Visualise the number of trips per hour and find the busiest hour



Remember, we took a fraction of trips. To find the actual number, you have to scale the number up by the sampling ratio.

**3.2.3** <font color = red>[2 mark]</font> <br>
Find the actual number of trips in the five busiest hours

In [ ]:
# Scale up the number of trips

# Fill in the value of your sampling fraction and use that to scale up the numbers
sample_fraction =



**3.2.4** <font color = red>[3 marks]</font> <br>
Compare hourly traffic pattern on weekdays. Also compare for weekend.

In [ ]:
# Compare traffic trends for the week days and weekends



What can you infer from the above patterns? How will finding busy and quiet hours for each day help us?

**3.2.5** <font color = red>[3 marks]</font> <br>
Identify top 10 zones with high hourly pickups. Do the same for hourly dropoffs. Show pickup and dropoff trends in these zones.

In [ ]:
# Find top 10 pickup and dropoff zones



**3.2.6** <font color = red>[3 marks]</font> <br>
Find the ratio of pickups and dropoffs in each zone. Display the 10 highest (pickup/drop) and 10 lowest (pickup/drop) ratios.

In [ ]:
# Find the top 10 and bottom 10 pickup/dropoff ratios



**3.2.7** <font color = red>[3 marks]</font> <br>
Identify zones with high pickup and dropoff traffic during night hours (11PM to 5AM)

In [ ]:
# During night hours (11pm to 5am) find the top 10 pickup and dropoff zones
# Note that the top zones should be of night hours and not the overall top zones



Now, let us find the revenue share for the night time hours and the day time hours. After this, we will move to deciding a pricing strategy.

**3.2.8** <font color = red>[2 marks]</font> <br>
Find the revenue share for nighttime and daytime hours.

In [ ]:
# Filter for night hours (11 PM to 5 AM)



##### Pricing Strategy

**3.2.9** <font color = red>[2 marks]</font> <br>
For the different passenger counts, find the average fare per mile per passenger.

For instance, suppose the average fare per mile for trips with 3 passengers is 3 USD/mile, then the fare per mile per passenger will be 1 USD/mile.

In [ ]:
# Analyse the fare per mile per passenger for different passenger counts




**3.2.10** <font color = red>[3 marks]</font> <br>
Find the average fare per mile by hours of the day and by days of the week

In [ ]:
# Compare the average fare per mile for different days and for different times of the day



**3.2.11** <font color = red>[3 marks]</font> <br>
Analyse the average fare per mile for the different vendors for different hours of the day

In [ ]:
# Compare fare per mile for different vendors



**3.2.12** <font color = red>[5 marks]</font> <br>
Compare the fare rates of the different vendors in a tiered fashion. Analyse the average fare per mile for distances upto 2 miles. Analyse the fare per mile for distances from 2 to 5 miles. And then for distances more than 5 miles.


In [ ]:
# Defining distance tiers



##### Customer Experience and Other Factors

**3.2.13** <font color = red>[5 marks]</font> <br>
Analyse average tip percentages based on trip distances, passenger counts and time of pickup. What factors lead to low tip percentages?

In [ ]:
#  Analyze tip percentages based on distances, passenger counts and pickup times



Additional analysis [optional]: Let's try comparing cases of low tips with cases of high tips to find out if we find a clear aspect that drives up the tipping behaviours

In [ ]:
# Compare trips with tip percentage < 10% to trips with tip percentage > 25%



**3.2.14** <font color = red>[3 marks]</font> <br>
Analyse the variation of passenger count across hours and days of the week.

In [ ]:
# See how passenger count varies across hours and days




**3.2.15** <font color = red>[2 marks]</font> <br>
Analyse the variation of passenger counts across zones

In [ ]:
# How does passenger count vary across zones



In [ ]:
# For a more detailed analysis, we can use the zones_with_trips GeoDataFrame
# Create a new column for the average passenger count in each zone.



Find out how often surcharges/extra charges are applied to understand their prevalance

**3.2.16** <font color = red>[5 marks]</font> <br>
Analyse the pickup/dropoff zones or times when extra charges are applied more frequently

In [ ]:
# How often is each surcharge applied?



## **4** Conclusion
<font color = red>[15 marks]</font> <br>

### **4.1** Final Insights and Recommendations
<font color = red>[15 marks]</font> <br>

Conclude your analyses here. Include all the outcomes you found based on the analysis.

Based on the insights, frame a concluding story explaining suitable parameters such as location, time of the day, day of the week etc. to be kept in mind while devising a strategy to meet customer demand and optimise supply.

**4.1.1** <font color = red>[5 marks]</font> <br>
Recommendations to optimize routing and dispatching based on demand patterns and operational inefficiencies

**4.1.2** <font color = red>[5 marks]</font> <br>

Suggestions on strategically positioning cabs across different zones to make best use of insights uncovered by analysing trip trends across time, days and months.

**4.1.3** <font color = red>[5 marks]</font> <br>
Propose data-driven adjustments to the pricing strategy to maximize revenue while maintaining competitive rates with other vendors.